google/cloud/forseti/scanner/scanners/iam_rules_scanner.py

```diff
@@ -32,6 +35,58 @@
 LOGGER = logger.get_logger(__name__)
 
 
+def _add_bucket_ancestor_bindings(policy_data):
+    """Add bucket relevant IAM policy bindings from ancestors.
+
+    Resources can inherit policy bindings from ancestors in the resource
+    manager tree. For example: a GCS bucket inherits a 'objectViewer' role
+    from a project or folder (up in the tree).
+
+    So far the IAM rules engine only checks the set of bindings directly
+    attached to a resource (direct bindings set (DBS)). We need to add
+    relevant bindings inherited from ancestors to DBS so that these are
+    also checked for violations.
+
+    If we find one more than one binding with the same role name, we need to
+    merge the members.
+
+    NOTA BENE: this function only handles buckets and bindings relevant to
+    these at present (but can and should be expanded to handle projects and
+    folders going forward).
+
+    Args:
+        policy_data (list): list of (parent resource, iam_policy resource,
+            policy bindings) tuples to find violations in.
+    """
+    storage_iam_roles = frozenset([
+        'roles/storage.admin',
+        'roles/storage.objectViewer',
+        'roles/storage.objectCreator',
+        'roles/storage.objectAdmin',
+    ])
+    bucket_data = [(r, bs) for (r, _, bs) in policy_data if r.type == 'bucket']
+    for bucket, bindings in bucket_data:
+        ancestor_bindings = [
+            bs for (r, _, bs) in policy_data
+            if r.full_name != bucket.full_name
+            and not bucket.full_name.find(r.full_name)]
+        for anbs in ancestor_bindings:
+            for anb in anbs:
```

Comment:
These nested for loops is really confusing, the previous list should be named as `all_ancestor_bindings`.
```
for ancestor_bindings in all_ancestor_bindings:
    for ancestor_binding in ancestor_bindings:

Reply:
Sorry for the confusion. Will do.